container_registry/container_analysis/samples.py

```diff
@@ -0,0 +1,343 @@
+        ValueError: If the parameters are invalid.
+    """
+    filterStr = "resourceUrl=\"" + image_url + "\""
+    client = container_analysis_client.ContainerAnalysisClient()
+    project_name = client.project_path(project_id)
+
+    response = client.list_occurrences(project_name, filter_=filterStr)
+    count = 0
+    for o in response:
+        # do something with the retrieved occurrence
+        # in this sample, we will simply count each one
+        count += 1
+    return count
+# [END occurrences_for_image]
+
+
+# [START pubsub]
+def pubsub(subscription_id, timeout, project_id):
+    """Handle incoming occurrences using a pubsub subscription
+
+    :param subscription_id: the user-specified identifier for the subscription
+    :param timeout: the number of seconds to listen for pubsub messages
+    :param project_id: the project's unique identifier
+    :return: number of occurrence pubsub messages received
+    """
+    client = SubscriberClient()
+    subscription_name = client.subscription_path(project_id, subscription_id)
+    receiver = MessageReceiver()
+    client.subscribe(subscription_name, receiver.pubsub_callback)
+
+    # listen for 'timeout' seconds
+    print("listening")
+    for _ in range(timeout):
+        sleep(1)
+    # print and return the number of pubsub messages received
+    print(receiver.msg_count)
+    return receiver.msg_count
+
+
+class MessageReceiver:
+    """Custom class to handle incoming pubsub messages
+    In this case, we will simply print and count each message as it comes in
+    """
+    def __init__(self):
+        # initialize counter to 0 on initialization
+        self.msg_count = 0
+
+    def pubsub_callback(self, message):
+        # every time a pubsub message comes in, print it and count it
+        self.msg_count += 1
+        print("Message " + str(self.msg_count) + ": " + message.data)
+        message.ack()
+
+
+def create_occurrence_subscription(subscription_id, project_id):
+    """Creates and returns a pubsub subscription listening to the
+    occurrence topic. This topic provides updates when occurrences are modified
+
+    :param subscription_id: the user-specified identifier for the subscription
+    :param project_id: the project's unique identifier
+    :return: a bool indicating whether the subscription is ready to use
+             if the subscription was already created, it will return True
+    :raises
+        google.api_core.exceptions.GoogleAPICallError: If the request
+            failed for any reason.
+        google.api_core.exceptions.RetryError: If the request failed due
+            to a retryable error and retry attempts failed.
+        ValueError: If the parameters are invalid.
+    """
+    topic_id = "resource-notes-occurrences-v1alpha1"
+    client = SubscriberClient()
+    topic_name = client.topic_path(project_id, topic_id)
+    subscription_name = client.subscription_path(project_id, subscription_id)
+    success = True
+    try:
```

Comment:
If possible, avoid exceptions / and try/catch as they may cause flow control to be a little less clear

Reply:
Here I'm just catching and responding to an exception thrown by the pubsub library, so it might be hard to avoid. I thought about doing get_subscription before create_subscription to explicitly check if it already exists first, but it looks like get_subscription throws an exception if the subscription doesn't exist too, so i'd have to try-catch either way. Is there a batter way to handle this I'm not thinking of?